draft-ietf-taps-interface.md

```diff
@@ -1784,37 +1775,183 @@ that the other endpoint is done sending on a connection.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the Framer that a new Connection is being created.
+Similarly, a stop event is delivered when a Connection is being torn down.
+The Framer can use the Connection object to look up specific properties
+of the connection or the network being used that may influence how
+to frame Messages.
+
+~~~
+MessageFramer -> Start(Connection)
+MessageFramer -> Stop(Connection)
+~~~
+
+When Message Framer receives a `Start` event, it has the opportunity to start
+writing some data on the Connection prior to the Connection delivering its
+`Ready` event. This allows the Framer to communicate control data to the
+remote endpoint that can be used to parse Messages.
+
+~~~
+MessageFramer.MakeConnectionReady(Connection)
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+MessageFramer.FailConnection(Connection, Error)
+~~~
+
+Before a Message Framer marks itself as ready, it can also dynamically add a protocol or framer
+above it in the stack. This allows protocols like STARTTLS, that need to add TLS conditionally,
+to modify the Protocol Stack based on a handshake result.
+
+~~~
+otherFramer := NewMessageFramer()
+MessageFramer.PrependFramer(Connection, otherFramer)
+~~~
+
+## Sender-side Message Framing {#send-framing}
+
+Message Framers deliver an event whenever a Connection sends a new Message.
+
+~~~
+MessageFramer -> NewSentMessage<Connection, MessageData, MessageContext, IsEndOfMessage>
+~~~
+
+Upon receiving this event, a Message Framer implementation is responsible for
+performing any necessary transformations and sending the resulting data to the next
+protocol. Implementations SHOULD ensure that there is a way to pass the original data
+through without copying to improve performance.
+
+~~~
+MessageFramer.Send(Connection, Data)
+~~~
+
+To provide an example, a simple protocol that adds a length as a header would receive
+the `NewSentMessage` event, create a data representation of the length of the Message
+data, and then send a block of data that is the concatenation of the length header and the original
+Message data.
+
+## Receiver-side Message Framing {#receive-framing}
+
+In order to parse an received flow of data into Messages, the Message Framer
+is first notified whenever new data is available to parse.
+
+~~~
+MessageFramer -> HandleReceivedData<Connection>
+~~~
+
+Upon receiving this event, a Message Framer can inspect the inbound data. The
+data is parsed from a particular cursor representing the unprocessed data. The
+Framer requests a specific amount of data it needs to have available in order to parse.
+If the data is not available, the parse fails.
+
+~~~
+MessageFramer.Parse(Connection, MinimumIncompleteLength, MaximumLength) -> (Data?, IsEndOfMessage?)
+~~~
+
```

Comment:
This implies a Framer would always operate on a byte stream / partial message.
Should we add an Interface to de-frame data that is already a Message?

Reply:
Hm, so it should be able to handle messages fine, since this is handling partial messages. I think the most common case of framer is over a bye stream, so I think handling that by default is important. But if you have a high `MinimumIncompleteLength`, any shorter Message will just be delivered along with `IsEndOfMessage`.

I suppose the thing to add here is the Message object itself.

Reply:
Added MessageContext here.